ui/components-react/src/components-react/filter-builder/FilterBuilderRuleGroup.tsx

```diff
@@ -0,0 +1,151 @@
+            {UiComponents.translate("filterBuilder.ruleGroup")}
+          </Button>
+        </ButtonGroup>
+      </div>
+    </div>
+  </div>;
+}
+
+/** @alpha */
+export interface PropertyFilterBuilderRuleGroupOperatorProps {
+  operator: PropertyFilterRuleGroupOperator;
+  onChange: (operator: PropertyFilterRuleGroupOperator) => void;
+}
+
+/** @alpha */
+export function PropertyFilterBuilderRuleGroupOperator(props: PropertyFilterBuilderRuleGroupOperatorProps) {
+  const {operator, onChange} = props;
+
+  const options = React.useMemo<Array<SelectOption<PropertyFilterRuleGroupOperator>>>(() => ([
+    { value: PropertyFilterRuleGroupOperator.And, label: UiComponents.translate("filterBuilder.operators.and") },
+    { value: PropertyFilterRuleGroupOperator.Or, label: UiComponents.translate("filterBuilder.operators.or") },
+  ]), []);
+
+  return <div className="rule-group-operator">
+    <Select options={options} value={operator} onChange={onChange} size="small" />
+  </div>;
+}
+interface PropertyFilterBuilderGroupOrRuleProps {
+  path: string[];
+  item: PropertyFilterBuilderRuleGroupItem;
+}
+
+function PropertyFilterBuilderGroupOrRule({path, item}: PropertyFilterBuilderGroupOrRuleProps) {
+  const itemPath = React.useMemo(() => ([...path, item.id]), [path, item]);
+
+  if (isPropertyFilterBuilderRuleGroup(item))
+    return <PropertyFilterBuilderRuleGroupRenderer path={itemPath} group={item} />;
+  return <PropertyFilterBuilderRuleRenderer path={itemPath} rule={item} />;
+}
+
+function useIsActive() {
+  const { isFocused, onBlur, onFocus } = useIsFocused();
+  const { isHovered, onMouseOver, onMouseOut } = useIsHovered();
+  const active = isHovered || isFocused;
+
+  return {
+    active,
+    eventHandlers: {
+      onMouseOver,
+      onMouseOut,
+      onFocus,
+      onBlur,
+    },
+  };
+}
+
+function useIsHovered() {
+  const [isHovered, setIsHovered] = React.useState(false);
+
+  const onMouseOver: React.MouseEventHandler<HTMLDivElement> = (e) => {
+    e.stopPropagation();
+    setIsHovered(true);
+  };
+
+  const onMouseOut: React.MouseEventHandler<HTMLDivElement> = (e) => {
+    e.stopPropagation();
+    setIsHovered(false);
+  };
+
+  return { isHovered, onMouseOver, onMouseOut };
+}
+
+function useIsFocused() {
+  const [isFocused, setIsFocused] = React.useState(false);
+  const timeout = React.useRef<any>();
+
+  React.useEffect(() => () => clearTimeout(timeout.current), []);
+
+  const onFocus: React.FocusEventHandler<HTMLDivElement> = (e) => {
+    e.stopPropagation();
+    clearTimeout(timeout.current);
+    timeout.current = undefined;
+    setIsFocused(true);
+  };
+
+  const onBlur: React.FocusEventHandler<HTMLDivElement> = (e) => {
+    e.stopPropagation();
+    // need to handle onBlur on the next tick to check if other child of element
+    // received onFocus event
+    timeout.current = setTimeout(() => {
+      setIsFocused(false);
+      timeout.current = undefined;
+    });
+  };
+
```

Comment:
The reason you need to resort to a hacky workaround is because each component hosts its own `isFocused` state. Such arrangement allows representing multiple focused components at the same time. The proper react way is to host a single `focusedRuleOrGroup` state which holds the element which is currently in focus.

While the current focus implementation works, there is little reason to change it now, other than to learn how to do things the react way.

Reply:
I was using react docs as an example: https://reactjs.org/docs/accessibility.html#mouse-and-pointer-events. I guess it might be an old example. I think if current implementation is good enough there is no need to spend more time on this and it can be reiterated later.

Reply:
The example on that documentation page puzzles me because it doesn't use [`FocusEvent.relatedTarget`](https://developer.mozilla.org/en-US/docs/Web/API/FocusEvent/relatedTarget) to detect which element has gained focus.